src/main/java/org/opensearch/flowframework/template/TemplateParser.java

```diff
@@ -0,0 +1,164 @@
+                throw new IllegalArgumentException("Edge connects node " + sourceNodeId + " to itself.");
+            }
+            edges.add(new ProcessSequenceEdge(sourceNodeId, destNodeId));
+        }
+
+        return topologicalSort(nodes, edges);
+    }
+
+    private static List<ProcessNode> topologicalSort(List<ProcessNode> nodes, List<ProcessSequenceEdge> edges) {
+        // Define the graph
+        Set<ProcessSequenceEdge> graph = new HashSet<>(edges);
+        // Map node id string to object
+        Map<String, ProcessNode> nodeMap = nodes.stream().collect(Collectors.toMap(ProcessNode::id, Function.identity()));
+        // Build predecessor and successor maps
+        Map<ProcessNode, Set<ProcessSequenceEdge>> predecessorEdges = new HashMap<>();
+        Map<ProcessNode, Set<ProcessSequenceEdge>> successorEdges = new HashMap<>();
+        for (ProcessSequenceEdge edge : edges) {
+            ProcessNode source = nodeMap.get(edge.getSource());
+            ProcessNode dest = nodeMap.get(edge.getDestination());
+            predecessorEdges.computeIfAbsent(dest, k -> new HashSet<>()).add(edge);
+            successorEdges.computeIfAbsent(source, k -> new HashSet<>()).add(edge);
+        }
+        // update predecessors on the node object
+        nodes.stream().filter(n -> predecessorEdges.containsKey(n)).forEach(n -> {
+            n.setPredecessors(predecessorEdges.get(n).stream().map(e -> nodeMap.get(e.getSource())).collect(Collectors.toSet()));
+        });
+
+        // See https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm
+        // L <- Empty list that will contain the sorted elements
+        List<ProcessNode> sortedNodes = new ArrayList<>();
+        // S <- Set of all nodes with no incoming edge
+        Queue<ProcessNode> sourceNodes = new ArrayDeque<>();
+        nodes.stream().filter(n -> !predecessorEdges.containsKey(n)).forEach(n -> sourceNodes.add(n));
+        if (sourceNodes.isEmpty()) {
+            throw new IllegalArgumentException("No start node detected: all nodes have a predecessor.");
+        }
+        logger.debug("Start node(s): {}", sourceNodes);
+
+        // while S is not empty do
+        while (!sourceNodes.isEmpty()) {
+            // remove a node n from S
+            ProcessNode n = sourceNodes.poll();
+            // add n to L
+            sortedNodes.add(n);
+            // for each node m with an edge e from n to m do
+            for (ProcessSequenceEdge e : successorEdges.getOrDefault(n, Collections.emptySet())) {
+                ProcessNode m = nodeMap.get(e.getDestination());
+                // remove edge e from the graph
+                graph.remove(e);
+                // if m has no other incoming edges then
+                if (!predecessorEdges.get(m).stream().anyMatch(i -> graph.contains(i))) {
+                    // insert m into S
+                    sourceNodes.add(m);
+                }
+            }
+        }
+        if (!graph.isEmpty()) {
```

Comment:
```suggestion
        if (!graph.isEmpty() || graph.size() != sortedNodes.size()) {
```

Reply:
I don't think this change is needed.  The only time I'd get to the second condition is if `graph.isEmpty()`.  In that case `graph.size()` is 0, and since we exited the above while loop, `sourceNodes.isEmpty()` and its size is therefor also 0, so this check ends up always being `0 != 0` which will never be true.